_posts/2020-08-02-prediction-of-medical-costs/2020-08-02-prediction-of-medical-costs.Rmd

```diff
@@ -0,0 +1,600 @@
+  terminator = terminator,
+  tuner = tuner_rs
+)
+
+# Train the model:
+at_ranger_final$store_tuning_instance = TRUE
+at_ranger_final$train(ins_task)
+```
+
+```{r eval=longrun, echo=FALSE}
+tab_tuning = at_ranger_final$archive$data()[, c("batch_nr", "regr.mse", "min.node.size", "mtry")]
+
+saveRDS(tab_tuning, file = "tab_tuning.rds")
+```
+
+```{r tabtuningrds, echo=FALSE, eval=!longrun}
+tab_tuning = readRDS("tab_tuning.rds")
+learner = lrn("regr.ranger", importance = "impurity")
+```
+
+From the tuning process of our final model, we conlude that the optimal hyperparameter setting is achieved through the combination of $min.node.size = 30$ and $mtry = 4$.
+The figure below illustrates the investigated parameter combinations via random search.
+The color intensity matches the value of the respective MSE and the hyperparameter combination with the best performance, which is used to train the model, is highlighted in red.  
+
+```{r echo=FALSE, eval=TRUE}
+ggplot() +
+  geom_point(aes(x = min.node.size, y = mtry, fill = regr.mse, col = regr.mse), data = tab_tuning, size = 5, shape = 21) +
+  geom_point(aes(x = min.node.size, y = mtry, fill = regr.mse), data = tab_tuning[regr.mse == min(tab_tuning$regr.mse), ], col = "red", size = 5, shape = 21, stroke = 2) +
+  annotate("text",
+    x = tab_tuning$min.node.size[tab_tuning$regr.mse == min(tab_tuning$regr.mse)],
+    y = tab_tuning$mtry[tab_tuning$regr.mse == min(tab_tuning$regr.mse)], col = "red", label = "min", vjust = -1) +
+  labs(title = "Visualization of Tuning Results", subtitle = "Random Search, Termination after 30 evaluations") +
+  theme_bw() +
+  labs(fill = "MSE", col = "MSE")
+```
+
+Eventually, we want to investigate how well the final model can explain the data observed. Overall, the model fit does not seem too bad: the majority of observations are close to the diagonal.
+There is, however, a considerable number of observations with charges higher than the ones predicted by the model.
+This could indicate that features are not considered with their true functionality or that the insurance company uses additional tariffication criteria outside the data set.
+
+```{r echo = FALSE, eval=longrun}
+# Compare predicted values to true values:
+png("prediction_at_ranger.png")
+autoplot(at_ranger_final$predict(ins_task)) +
+  theme_bw()
+dev.off()
+```
+
+![Prediction autotuned ranger](prediction_at_ranger.png)
+
+## Feature Importance
+
+As a last step of our analysis, we want to shed some light onto the functionality of our final "black box" model.
+To get a better understanding of the predictions made, we investigate the importance of each individual covariate.
+This can be done with the help of filter methods, which assign an importance value to each feature, which will then be ranked.
+For the Ranger learning algorithm, we can make use of the filter methods embedded in the mlr3-package.  
+
+```{r}
+filter = flt("importance", learner = learner)
+filter$calculate(ins_task)
+```
+
```

Comment:
Mention: Note, that the feature importance computed here does not necessarily reflect the feature importance in the model trained above, as different sub-sets of the data and hyperparameters can influence the computed importance scores. 
A full set of post-hoc tools for post-hoc model interpretation for mlr3 models can be found in the iml and DALEX packages, (link to packages), the mlr3 book contains a section on how to use those packages together with mlr3. (Link to book).